### PR TITLE
docs: Update README

### LINKED ISSUE #16: Zig PDK README

### PR DESCRIPTION
**Pull request by mhmd-azeez**
- [x] Write a similar README to go-pdk https://github.com/extism/go-pdk
- [x] Change the API to make it more convenient
   - [x] Move `allocate`, `allocateBytes`, `findMemory` to `Plugin`. Since `Memory` is not public, there was no way to call them
   - [x] Added `setError` and `setErrorMemory`
   - [x] Added `getVarInt` and `setVarInt`
- [x] fix build errors

Fixes #16

### CODE REVIEW

**Comment by mhmd-azeez** [comment]
Not sure why there were some build errors on my machine

**Comment by nilslice** [comment]
@mhmd-azeez - thank you! which version of Zig are you using with this? 

I'm on `0.12.0-dev.978+78855bd21`, and seem to need to use `std.builtin.Endian.Little` (vs. `std.builtin.Endian.little`, note the capitalization of `little`). I am not sure if this is a version thing or a typo haha

**Comment by mhmd-azeez** [comment]
@nilslice seems like they have changed it to `little` [in the newer versions](https://github.com/ziglang/zig/blob/master/lib/std/builtin.zig#L454):
```
PS D:\dylibso\zig-pdk> zig build
zig build-exe Basic example Debug wasm32-freestanding-musl: error: the following command failed with 2 compilation errors:
D:\tools\zig.exe build-exe D:\dylibso\zig-pdk\examples\basic.zig --cache-dir D:\dylibso\zig-pdk\zig-cache --global-cache-dir C:\Users\muham\AppData\Local\zig --name Basic example -rdynamic -target wasm32-freestanding-musl -mcpu generic --mod extism-pdk::D:\dylibso\zig-pdk\src\main.zig --deps extism-pdk --listen=-
Build Summary: 0/3 steps succeeded; 1 failed (disable with --summary none)
install transitive failure
└─ install Basic example transitive failure
   └─ zig build-exe Basic example Debug wasm32-freestanding-musl 2 errors
src\main.zig:31:73: error: enum 'builtin.Endian' has no member named 'Little'
            std.mem.writeInt(u64, buf[i..][0..8], x, std.builtin.Endian.Little);
                                                                        ^~~~~~
D:\tools\lib\std\builtin.zig:452:20: note: enum declared here
pub const Endian = enum {
                   ^~~~
referenced by:
    count_vowels: examples\basic.zig:19:25
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
src\Memory.zig:42:78: error: enum 'builtin.Endian' has no member named 'Little'
        const data = std.mem.readInt(u64, buf[i..][0..8], std.builtin.Endian.Little);
                                                                             ^~~~~~
D:\tools\lib\std\builtin.zig:452:20: note: enum declared here
pub const Endian = enum {
                   ^~~~
PS D:\dylibso\zig-pdk> zig version
0.12.0-dev.1664+8ca4a5240
```

**Comment by mhmd-azeez** [comment]
Hah! it's been changed 3 weeks ago: https://github.com/ziglang/zig/commit/3fc6fc68129219a026ae3d7dff82513758e33a21

**Comment by nilslice** [comment]
haha! thank you for looking at that. sorry I was behind a few versions... I think our strategy has been to follow the latest versions as close as possible. 

@usdogu can advise, as the expert Zig extism contributor!

**Comment by mhmd-azeez** [comment]
In addition to reviewing the API changes, please also take a close look at the examples to make sure allocations and deallocations are properly managed

**Comment by bhelx** [comment]
Thanks for taking this one on!

File: src/main.zig [hunk 1]
```diff
@@ -99,32 +139,48 @@ pub const Plugin = struct {
         return value;
     }
 
+    pub fn getVarInt(self: Plugin, comptime T: type, key: []const u8) !?T {
+        const result = try self.getVar(key);
+
+        if (result) |buf| {
+            return std.mem.readPackedInt(i32, buf, 0, .little);
```
Comment (usdogu) [inline src/main.zig]:
```suggestion
            return std.mem.readPackedInt(T, buf, 0, .little);
```

File: src/main.zig [hunk 2]
```diff
@@ -99,32 +139,48 @@ pub const Plugin = struct {
...
+
+        self.setVar(key, buffer);
+    }
+
     pub fn removeVar(self: Plugin, key: []const u8) void {
-        _ = self; // to make the interface consistent
-        const mem = Memory.allocateBytes(key);
+        const mem = self.allocateBytes(key);
         defer mem.free();
         extism.extism_var_set(mem.offset, 0);
```
Comment (usdogu) [inline src/main.zig]:
```suggestion
        extism.var_set(mem.offset, 0);
```

File: src/main.zig [hunk 3]
```diff
@@ -99,32 +139,48 @@ pub const Plugin = struct {
...
+        const key_mem = self.allocateBytes(key);
         defer key_mem.free();
-        const val_mem = Memory.allocateBytes(value);
+        const val_mem = self.allocateBytes(value);
         defer val_mem.free();
         extism.var_set(key_mem.offset, val_mem.offset);
     }
 
+    pub fn setVarInt(self: Plugin, comptime T: type, key: []const u8, value: T) !void {
+        const buffer = try self.allocator.alloc(u8, 8);
```
Comment (usdogu) [inline src/main.zig]:
```suggestion
        const buffer = try self.allocator.alloc(u8, @sizeOf(T));
```

File: src/main.zig [hunk 4]
```diff
@@ -99,32 +139,48 @@ pub const Plugin = struct {
...
+        errdefer self.allocator.free(buffer);
```
Comment (usdogu) [inline src/main.zig]:
```suggestion
        defer self.allocator.free(buffer);
```

**Review by usdogu — CHANGES_REQUESTED** [review]
Other than these, LGTM!

File: src/main.zig (hunk 1 above)
Comment (mhmd-azeez) [reply src/main.zig]: Great catch!

**Comment by mhmd-azeez** [comment]
@usdogu thank you very much for you suggestions!

**Comment by mhmd-azeez** [comment]
Also added the readme examples to `basic.zig`, we can turn them into integration tests

**Review by usdogu — APPROVED** [review]

**Comment by nilslice** [comment]
Going to merge this one & make sure #19 works based on it

**Pull request merged**